src/main/java/pl/edu/agh/firecell/engine/algorithm/BasicAlgorithm.java

```diff
@@ -1,52 +1,112 @@
+        boolean newFlammable = oldCell.flammable();
+        int newBurningTime = oldCell.burningTime();
+        Material newMaterial = oldCell.material();
 
-        double xTemp = computeConductivity(
-                oldState.getCell(IndexUtils.east(cellIndex)),
-                oldCell,
-                oldState.getCell(IndexUtils.west(cellIndex)));
+        //computeFirePropagation();
+        //computeSmokePropagation();
 
-        double averageTemp = DoubleStream.of(yTemp, zTemp, xTemp)
-                .average()
-                .orElse(oldCell.temperature());
+        return new Cell(
+                newTemperature,
+                newBurningTime,
+                newFlammable,
+                newMaterial
+        );
+    }
 
-        return new Cell(averageTemp, oldCell.conductivityCoefficient(), 0, false, oldCell.material());
+    private double computeNewTemperature(State oldState, Vector3i cellIndex, Cell oldCell) {
+        return switch (oldCell.material().getMatterState()) {
+            case SOLID -> oldCell.temperature() + computeConductivityFromAll(oldState, oldCell, cellIndex);
+            case FLUID -> oldCell.temperature() + computeConvectionForFluid(oldState, oldCell, cellIndex);
+        };
     }
 
     private double computeConductivity(Cell former, Cell middle, Cell latter) {
+        return deltaTime * (
+                GAMMA_PRIM_N1 * (middle.temperature() - former.temperature()) -
+                        GAMMA_PRIM_N1 * (latter.temperature() - middle.temperature())
+        );
+    }
+
+    private double computeConductivityFromAll(State oldState, Cell oldCell, Vector3i cellIndex) {
+
+        double yTemp;
+        double zTemp;
+        double xTemp;
+
+        try {
 
-        // TODO: put them into class constants and give meaningful names
-        double dt = 0.05;
-        double gammaPrimN = 0;
-        double gammaPrimN1 = 0;
+            yTemp = computeConductivity(
+                    oldState.getCell(IndexUtils.north(cellIndex)),
+                    oldCell,
+                    oldState.getCell(IndexUtils.south(cellIndex)));
 
-        return middle.temperature() +
-                dt * (
-                        gammaPrimN * (middle.temperature() - former.temperature()) -
-                                gammaPrimN1 * (latter.temperature() - middle.temperature())
-                );
+            zTemp = computeConductivity(
+                    oldState.getCell(IndexUtils.up(cellIndex)),
+                    oldCell,
+                    oldState.getCell(IndexUtils.down(cellIndex)));
+
+            xTemp = computeConductivity(
+                    oldState.getCell(IndexUtils.east(cellIndex)),
+                    oldCell,
+                    oldState.getCell(IndexUtils.west(cellIndex)));
+
+        } catch (IndexOutOfBoundsException e) {
+            logger.debug(String.valueOf(e));
+            return oldCell.temperature();
```

Comment:
It's wrong that we discard all conductivity from all neighbours from all directions only because at least one of them doe not exist.
But for now let's leave it.
After merge create a Jira task for this.